[ Sonoma Debug arm64]: ASSERTION FAILED: m_clients.contains(client) in imported/w3c/web-platform-tests/media-source/URL-createObjectURL-null.html

jyavenard opened the pull request:
#### 9a18e435893e9e063fcfa1670952c0e9eedfe76f
<pre>
[ Sonoma Debug arm64]: ASSERTION FAILED: m_clients.contains(client) in imported/w3c/web-platform-tests/media-source/URL-createObjectURL-null.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=278659">https://bugs.webkit.org/show_bug.cgi?id=278659</a>
<a href="https://rdar.apple.com/134709579">rdar://134709579</a>

Reviewed by Youenn Fablet.

This has been a long standing issue.
When the HTMLMediaElement was being destructed, it would call MediaSource::detachElement.
MediaSource::detachElement then goes through the official steps of detaching
all source buffers, which then detach the various tracks from the media element.
However, by that time the HTMLMediaElement is already mid-destruction. Its tracks
have already been detached.
Attempting to detach the tracks again would cause the assertion.

To disambiguate between when the MediaSource is being detached from the element
when the src/srcObject are being cleared from when the HTMLMediaElement is being
destructed, we add a new MediaSource::elementIsShuttingDown method which will
clear the MediaSource&apos;s link to the HTMLMediaElement. This will stop the MediaSource
from making re-entrant call into the media element while it&apos;s being destroyed.

Covered by existing tests.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::elementIsShuttingDown):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.cpp:
(WebCore::MediaSourceInterfaceMainThread::elementIsShuttingDown):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.h:
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceProxy.h:
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.cpp:
(WebCore::MediaSourceInterfaceWorker::elementIsShuttingDown):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::~HTMLMediaElement):

Canonical link: <a href="https://commits.webkit.org/285110@main">https://commits.webkit.org/285110@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3df97da1300806431faa623c6041f48ca89a433d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71585 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50998 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24359 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75697 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22790 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73700 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58799 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22610 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15013 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74651 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46269 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61661 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36990 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42932 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19127 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21131 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64836 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19491 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77415 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15819 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18668 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64253 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15862 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/61696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64250 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15813 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12396 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6035 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46798 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/1577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47869 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49153 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47611 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/HTMLMediaElement.h:
```diff
@@ -336,7 +336,7 @@ class HTMLMediaElement
 #if ENABLE(MEDIA_SOURCE)
 //  Media Source.
-    void detachMediaSource();
+    void detachMediaSource(bool shuttingDown = false);
```
Boolean parameters like this are frowned upon in WebKit because they lead to poor readability at call sites. This should use an enum class.

rniwa replied:
maybe we should add it to coding guidelines?

cdumez replied:
https://webkit.org/code-style-guidelines/#names-enum-to-bool

jyavenard replied:
is it okay to have default parameter value still ?

cdumez replied:
> is it okay to have default parameter value still ?

I think so?

cdumez commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -734,7 +734,7 @@ HTMLMediaElement::~HTMLMediaElement()
 
 #if ENABLE(MEDIA_SOURCE)
-    detachMediaSource();
+    detachMediaSource(true);
```
Note how we have no idea what `true` means here..

jyavenard replied:
I totally agree. I shouldn't have pushed this change as-is.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -1220,6 +1220,14 @@ bool MediaSource::isEnded() const
     return readyState() == ReadyState::Ended;
 }
 
+void MediaSource::elementShuttingDown()
```
```suggestion
void MediaSource::elementIsShuttingDown()
```

youennf commented on Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.cpp:
```diff
@@ -111,6 +111,14 @@ void MediaSourceInterfaceWorker::detachFromElement()
+void MediaSourceInterfaceWorker::elementShuttingDown()
+{
+    ASSERT(m_handle->hasEverBeenAssignedAsSrcObject());
+    m_handle->ensureOnDispatcher([](MediaSource& mediaSource) {
```
auto&?

jyavenard committed:
[ Sonoma Debug arm64]: ASSERTION FAILED: m_clients.contains(client) in imported/w3c/web-platform-tests/media-source/URL-createObjectURL-null.html
https://bugs.webkit.org/show_bug.cgi?id=278659
rdar://134709579

Reviewed by Youenn Fablet.

This has been a long standing issue.
When the HTMLMediaElement was being destructed, it would call MediaSource::detachElement.
MediaSource::detachElement then goes through the official steps of detaching
all source buffers, which then detach the various tracks from the media element.
However, by that time the HTMLMediaElement is already mid-destruction. Its tracks
have already been detached.
Attempting to detach the tracks again would cause the assertion.

To disambiguate between when the MediaSource is being detached from the element
when the src/srcObject are being cleared from when the HTMLMediaElement is being
destructed, we add a new MediaSource::elementIsShuttingDown method which will
clear the MediaSource's link to the HTMLMediaElement. This will stop the MediaSource
from making re-entrant call into the media element while it's being destroyed.

Covered by existing tests.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::elementIsShuttingDown):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.cpp:
(WebCore::MediaSourceInterfaceMainThread::elementIsShuttingDown):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceMainThread.h:
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceProxy.h:
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.cpp:
(WebCore::MediaSourceInterfaceWorker::elementIsShuttingDown):
* Source/WebCore/Modules/mediasource/MediaSourceInterfaceWorker.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::~HTMLMediaElement):

Canonical link: https://commits.webkit.org/285110@main
